### PR TITLE
Feature/reservoirs mvp

### PR DESCRIPTION
**Pull request by westb2**
Not yet ready for review! Making several final edits and going through commenting pull request for clarity. Will notify when ready.

### CODE REVIEW

File: docker/dev/Dockerfile
Comment (westb2) [inline docker/dev/Dockerfile:1]: I would like my docker workflow in the main branch because I believe we should have an image to do dev from, however given that the release image inherits some stuff from here and I want this image to build parflow with the debugger enabled I believe maybe we need a third image. Figured I would open up the disussion.

**Comment by westb2** [comment]
Just realized I will need to figure out how to make reservoirs.names optional or this breaks our test suite

**Comment by westb2** [comment]
Also, should I check in a test case with this for reservoirs?

**Comment by smithsg84** [comment]
Yes, we should have a test case for features.

**Comment by westb2** [comment]
Will be uploading an appropriate test case from my set of personal test cases later, then this should be ready

**Comment by westb2** [comment]
Pending workflow success this PR is ready for review. Since last time, I have added code to support multiple mpi ranks. I have also added two test cases, one for reservoirs as is and one to confirm the work when parallel processing.

**Comment by westb2** [comment]
Just realized at the time of my last comment a push had failed. All changes are in now.

File: pftools/python/parflow/tools/builders.py [hunk 1]
```diff
@@ -619,9 +633,9 @@ def apply(self, run=None, name_registration=True):
         if name_registration:
             names_to_set = addon_keys
             for unit_name in valid_unit_names:
-                if unit_name in self.name_registration:
+                if unit_name.casefold() in (registered_name.casefold() for registered_name in self.name_registration):
```
Comment (smithsg84) [inline pftools/python/parflow/tools/builders.py:636]: Why is case being removed?   We don't traditionally do that in PF.

File: test/python/CMakeLists.txt [hunk 2]
```diff
@@ -25,6 +25,8 @@ if(${PARFLOW_HAVE_HYPRE})
     LW_var_dz_spinup
     terrain_following_grid_overland
     var_dz_1D
+    reservoir_mpi_test
```
Comment (smithsg84) [inline test/python/CMakeLists.txt:28]: This doesn't look to be tested in parallel; doesn't look like it is testing under MPI.  Needs to be in the PARALLEL_ topo tests.

File: test/python/reservoir_mpi_test.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+from parflow.tools.fs import mkdir, get_absolute_path
+import pandas as pd
+
+sloping_slab = Run("model", __file__)
+
+#---------------------------------------------------------
+
+sloping_slab.FileVersion = 4
+
+sloping_slab.Process.Topology.P = 1
```
Comment (smithsg84) [inline test/python/reservoir_mpi_test.py:17]: Looks to be hard-coding to a single rank run?

File: pftools/python/parflow/tools/io.py [hunk 4]
```diff
@@ -390,7 +390,7 @@ def __init__(
         read_sg_info: bool=False
     ):
         self.filename = file
-        self.f = open(self.filename, 'rb')
+        self.f = open(self.filename, 'rb+')
```
Comment (smithsg84) [inline pftools/python/parflow/tools/io.py:393]: This is a Reader why is writing being added?

File: pfsimulator/parflow_lib/reservoir_package.c [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+      amps_Invoice release_cell_invoice = amps_NewInvoice("%i", &release_cell_rank);
+      amps_AllReduce(amps_CommWorld, release_cell_invoice, amps_Max);
+      amps_FreeInvoice(release_cell_invoice);
+
+      part_of_reservoir_lives_on_this_rank =
+              (release_cell_rank==current_mpi_rank) ||
+              (intake_cell_rank==current_mpi_rank) ||
+              (secondary_intake_cell_rank==current_mpi_rank);
+
+      MPI_Comm new_reservoir_communicator;
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/reservoir_package.c:285]: Will break if MPI is not used.   Need to guard MPI usage.

#ifdef PARFLOW_HAVE_MPI
#else
#endif

The AMPS layer was supposed to hide MPI and do nothing/appropriate when MPI or other parallel layers where not present.   So for things that are not exposed by AMPS we need to guard appropriately or expand the AMPS layer to add the functionality.

File: pfsimulator/parflow_lib/reservoir_package.c [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+    double secondary_intake_x_location, secondary_intake_y_location;
+    double release_x_location, release_y_location;
+    double max_storage, min_release_storage, storage, release_rate;
+} Type0;                      /* basic vertical reservoir */
+
+/*--------------------------------------------------------------------------
+ * ReservoirPackage
+ *--------------------------------------------------------------------------*/
+
+bool subgrid_lives_on_this_rank(Subgrid* subgrid, Grid *grid){
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/reservoir_package.c:124]: New code should have doxygen comments for methods.

https://github.com/parflow/parflow/blob/master/CONTRIBUTING.md#source-code-documentation

File: pfsimulator/parflow_lib/reservoir_package.c [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+      amps_FreeInvoice(release_cell_invoice);
+
+      part_of_reservoir_lives_on_this_rank =
+              (release_cell_rank==current_mpi_rank) ||
+              (intake_cell_rank==current_mpi_rank) ||
+              (secondary_intake_cell_rank==current_mpi_rank);
+
+      MPI_Comm new_reservoir_communicator;
+      split_color = part_of_reservoir_lives_on_this_rank ? 1 : MPI_UNDEFINED;
+      MPI_Comm_split(MPI_COMM_WORLD, split_color, current_mpi_rank, &new_reservoir_communicator);
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/reservoir_package.c:287]: Should use amps_CommWorld for MPI_COMM_WORLD.

File: pfsimulator/parflow_lib/solver_richards.c [hunk 8]
```diff
@@ -1214,7 +1214,10 @@ SetupRichards(PFModule * this_module)
     /*-------------------------------------------------------------------
      * Print out the initial well data?
      *-------------------------------------------------------------------*/
-
+    WriteReservoirs("ReservoirsOutput",
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/solver_richards.c:1217]: Add a key to control writing the output, like the print_wells flag.   We allow users to turn off I/O if they don't want it.

File: pfsimulator/parflow_lib/reservoir.h [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+#define FLUX_STANDARD  0
+#define FLUX_WEIGHTED  1
+#define FLUX_PATTERNED 2
+
+/*     Write Options      */
+#define RESERVOIRDATA_DONTWRITEHEADER 0
+#define RESERVOIRDATA_WRITEHEADER     1
+
+/*     Print Flags      */
+#define RESERVOIRDATA_PRINTPHYSICAL  0x0001
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/reservoir.h:212]: Are these flags used somewhere?   I'm not seeing it show up in search.

File: pfsimulator/parflow_lib/reservoir.h [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    double release_x_lower, release_y_lower, release_z_lower;
+    double release_x_upper, release_y_upper, release_z_upper;
+    double diameter;
+    double max_storage, min_release_storage, storage, release_rate;
+    double intake_amount_since_last_print, release_amount_since_last_print, release_amount_in_solver;
+    Subgrid       *intake_subgrid;
+    Subgrid       *secondary_intake_subgrid;
+    Subgrid       *release_subgrid;
+    double size;
+    MPI_Comm mpi_communicator;
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/reservoir.h:57]: MPI_Comm needs to be guarded, may not have MPI in build.

File: pfsimulator/parflow_lib/solver_richards.c [hunk 11]
```diff
@@ -3027,7 +3030,130 @@ AdvanceRichards(PFModule * this_module, double start_time,      /* Starting time
...
+                                 flux_in += pp_sp[ip] * dx * dy;
+                                 pp_sp[ip] = reservoir_reset_pressure;
+                               }
+                             }
+                           }
+              );
+            }
+          }
+        }
+        if (ReservoirDataPhysicalMpiCommunicator(reservoir_data_physical) != MPI_COMM_NULL){
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/solver_richards.c:3145]: This will break if we build without MPI.

**Review by smithsg84 — CHANGES_REQUESTED** [review]
Some cleanup fixes and there looks like an issue with multiple processors appending to a shared file.

**Comment by smithsg84** [comment]
Test build is failing; tests require the C to be clean.   Should not show any errors when building with 

-DCMAKE_C_FLAGS="-Wall -Werror"

/home/runner/work/parflow/parflow/pfsimulator/parflow_lib/problem_phase_source.c:74:60: error: unused variable ‘reservoir_intake_subgrid’ [-Werror=unused-variable]
   74 |   Subgrid          *subgrid, *well_subgrid, *tmp_subgrid, *reservoir_intake_subgrid, *reservoir_release_subgrid;

**Comment by westb2** [comment]
Thanks for the review Steve. Sorry to make you leave the Doxygen comment on two in a row - I should have gone back and updated this one after your var-dz review. I've addressed most of your comments but am going to dig a little deeper into the safety of printing that file and ensure the build is passing. I will leave another comment once I have done both those things

**Comment by westb2** [comment]
This is ready for review again when you get the time @smithsg84. I think the tests should be passing as well but I don't have permission to run them here. Since you last looked I have made a minor change outside of addressing your comments. Specifically, I added code within parflow to edit the slopes at the reservoir intake cells such that flow would stop. I have functions that I think work for both overland kinematic and overland flow boundary conditions. However, I am actually a bit unclear on what the best way would be to choose which one to do. It seems we might allow people to use both conditions on the same domain. The one place I can find that seems to be doing what I want is in bc_pressure_package.c, where the conditions are set up. Should I do a loop similar to that to check what type of surface each reservoir is on? Thanks in advance.

**Comment by smithsg84** [comment]
@westb2 , PF requires code to compile cleanly to merge.   There are errors showing up in the CI tests because of this.   Using 

 -DCMAKE_C_FLAGS="-Wall -Werror -Wno-unused-result -Wno-unused-function"
 
 Will flag the errors.  This is useful because the warnings do show real errors, like the code in reservoir_package.c :
```
void stop_outlet_flow_at_cell_overland_flow(int i, int j, ProblemData* problem_data, Grid* grid){
  Vector      *slope_x = ProblemDataTSlopeX(problem_data);
  Vector      *slope_y = ProblemDataTSlopeY(problem_data);
  Subvector   *slope_x_subvector ;
  Subvector   *slope_y_subvector;
  int index_slope_x;
  int index_slope_y;
  double *slope_x_ptr = SubvectorData(slope_x_subvector);;
  double *slope_y_ptr = SubvectorData(slope_y_subvector);
```

the subvectors are being used without being initialized.

**Comment by westb2** [comment]
I think those changes must have been fatfingered in my last intermediate commit, or my test cases never would have passed. Regardless, I checked in a fix as well as added a test case for the overland flow boundary condition. Stuff on my end is building, and my tests are passing. So I think this is ready for review, unless somehow tests are failing in workflow

**Comment by westb2** [comment]
As a last update, we now have a key that is required if any reservoirs are in the domain of "Reservoirs.Overland_Flow_Solver" to indicate what solver is being used in the domain. It will throw an error if an unsupported option is provided. Laura and I felt it was the fastest fix that would not require handling weird edge cases or implementing an overcomplicated solution where we looped over the patches

**Comment by smithsg84** [comment]
Compile is still failing.

**Comment by westb2** [comment]
I am not sure why, but I don't see that error on my local build even with that cmake flag. Not sure why, feels like an unused variable should be a system agnostic warning. I will resolve it, but is there a way to trigger the github build on my own? Don't want to keep bugging you for minor stuff like this.

**Comment by smithsg84** [comment]
Are GitHub actions not running if you create a branch?   When I fork the repo and create a branch, pushes to the branch trigger the CI to run.   I thought if you owned the repo you could trigger actions on it.  

https://github.com/westb2/parflow/actions

The merge checks are run as part of the project and the setup was configured to run require a project owner to trigger builds for first time contributors.   Once a contribution is approved, future PR should not require actions.  GitHub setup things this way since people were doing PR's that ran bitcoin like stuff and chewing up compute time.

**Comment by smithsg84** [comment]
Warning and errors can be compiler specific.   The Ubuntu 22.04 build is using GCC GNU 11.4.0.

**Comment by westb2** [comment]
If not it's fine as I need to figure out how to get those warnings showing on my local anyways, but it would be nice to confirm that the build isn't failing for some other reason on remote before asking for more of your time.

**Comment by westb2** [comment]
Hi @smithsg84. Just went through and made sure I had addressed all your changes. There was one comment on a test I wasn't sure what our pattern was (regarding hardcoding number of ranks), but other than that whenever you get a chance to review this again it is ready for you.

File: pfsimulator/parflow_lib/reservoir_package.c [hunk 12]
```diff
@@ -0,0 +1,594 @@
...
+      iz_lower = grid_nz - 1;
+
+      nx = 1;
+      ny = 1;
+      nz = 1;
+
+      rx = 0;
+      ry = 0;
+      rz = 0;
+      #ifdef PARFLOW_HAVE_MPI
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/reservoir_package.c:294]: General comment , we don't indent preprocessor statements.

File: pfsimulator/parflow_lib/solver_richards.c [hunk 13]
```diff
@@ -3031,7 +3043,129 @@ AdvanceRichards(PFModule * this_module, double start_time,      /* Starting time
...
+            nz = SubgridNZ(tmp_subgrid);
+            double dx = SubgridDX(tmp_subgrid);
+            double dy = SubgridDY(tmp_subgrid);
+            int grid_nz = SubgridNZ(subgrid);
+
+            pp_sp = SubvectorData(p_sub_sp);
+            GrGeomInLoop(i, j, k, gr_domain, r, ix, iy, iz, nx, ny, nz,
+                         {
+                           ip = SubvectorEltIndex(p_sub_sp, i, j, k);
+                           if (k == (grid_nz - 1))
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/solver_richards.c:3100]: This appears to be an assumption about where top of the domain is?

File: pfsimulator/parflow_lib/solver_richards.c [hunk 14]
```diff
@@ -3031,7 +3043,129 @@ AdvanceRichards(PFModule * this_module, double start_time,      /* Starting time
...
+              nz = SubgridNZ(tmp_subgrid);
+              double dx = SubgridDX(tmp_subgrid);
+              double dy = SubgridDY(tmp_subgrid);
+              int grid_nz = SubgridNZ(subgrid);
+
+              pp_sp = SubvectorData(p_sub_sp);
+              GrGeomInLoop(i, j, k, gr_domain, r, ix, iy, iz, nx, ny, nz,
+                           {
+                             ip = SubvectorEltIndex(p_sub_sp, i, j, k);
+                             if (k == (grid_nz - 1))
```
Comment (smithsg84) [inline pfsimulator/parflow_lib/solver_richards.c:3139]: Assumption about where top of domain is?

**Review by smithsg84 — CHANGES_REQUESTED** [review]
Did a more careful review and found some other issues.

Is there an assumption about where the domain top is?   It looks like there is an assumption that the top of domain is at the top of the grid    grid_nx - 1?   Maybe I'm confused.

**Comment by westb2** [comment]
Hi Steve. Thanks for another careful review. I believe I have addressed all of your comments

**Review by smithsg84 — APPROVED** [review]
